src/types.h

```diff
@@ -101,8 +101,8 @@ class SizedType
 private:
   size_t size_; // in bytes
   bool is_signed_ = false;
-  SizedType *element_type_ = nullptr; // for "container" and pointer
-                                      // (like) types
+  std::shared_ptr<SizedType> element_type_; // for "container" and pointer
```

Comment:
Why not unique_ptr? We're not sharing ownership anywhere.

Reply:
We do a lot of copies of types everywhere (see snippet below). For most types it doesn't really matter but for array/pointer types we need to recursively copy the type. The hacky fix for now is to use a sharedptr to manage the ownership mess for us. 



```
/vagrant/src/ast/semantic_analyser.cpp: In member function ‘virtual void bpftrace::ast::SemanticAnalyser::visit(bpftrace::ast::AssignVarStatement&)’:
/vagrant/src/ast/semantic_analyser.cpp:2099:43: error: use of deleted function ‘bpftrace::SizedType& bpftrace::SizedType::operator=(const bpftrace::SizedType&)’
 2099 |   assignment.var->type = assignment.expr->type;
      |                                           ^~~~
/vagrant/src/ast/semantic_analyser.cpp:2114:43: error: use of deleted function ‘bpftrace::SizedType& bpftrace::SizedType::operator=(const bpftrace::SizedType&)’
 2114 |         search->second = assignment.expr->type;
      |                                           ^~~~
/vagrant/src/ast/semantic_analyser.cpp:2128:49: error: use of deleted function ‘bpftrace::SizedType& bpftrace::SizedType::operator=(const bpftrace::SizedType&)’
 2128 |     variable_val_[var_ident] = assignment.expr->type;
      |                                                 ^~~~
/vagrant/src/ast/semantic_analyser.cpp:2129:45: error: use of deleted function ‘bpftrace::SizedType& bpftrace::SizedType::operator=(const bpftrace::SizedType&)’
 2129 |     assignment.var->type = assignment.expr->type;
      |                                             ^~~~
/vagrant/src/ast/semantic_analyser.cpp:2175:23: error: use of deleted function ‘bpftrace::SizedType::SizedType(const bpftrace::SizedType&)’
 2175 |       auto var_type = storedTy;
      |                       ^~~~~~~~
/vagrant/src/ast/semantic_analyser.cpp:2176:24: error: use of deleted function ‘bpftrace::SizedType::SizedType(const bpftrace::SizedType&)’
 2176 |       auto expr_type = assignTy;
      |                        ^~~~~~~~
/vagrant/src/ast/semantic_analyser.cpp: In member function ‘void bpftrace::ast::SemanticAnalyser::assign_map_type(const bpftrace::ast::Map&, const bpftrace::SizedType&)’:
/vagrant/src/ast/semantic_analyser.cpp:2821:26: error: use of deleted function ‘bpftrace::SizedType& bpftrace::SizedType::operator=(const bpftrace::SizedType&)’
 2821 |         search->second = type;
```

Reply:
Ah, I see. I guess this PR is the simplest fix for now.